src/python/pants/util/strutil.py

```diff
@@ -335,3 +342,35 @@ def wrapper(func: Callable[P, R]) -> Callable[P, R]:
         return func
 
     return wrapper
+
+
+class _JsonEncoder(json.JSONEncoder):
+    """Allow us to serialize everything, with a fallback on `str()` in case of any esoteric
+    types."""
+
+    def default(self, o):
+        """Return a serializable object for o."""
+        if dataclasses.is_dataclass(o):
+            return dataclasses.asdict(o)
```

Comment:
One of the other comments mentioned including the dataclass name. What about doing it like this?

```suggestion
            return {o.__class__.__name__: dataclasses.asdict(o)}
```

Reply:
I think the content is the same either way. Whatever "name" we give the content doesn't have much meaning.

Reply:
Oh? Isn't the cache key type aware? Types are significant for scheduling the rules, I assumed they were also relevant for caching.

Reply:
> I think the content is the same either way. Whatever "name" we give the content doesn't have much meaning.

The content is the same, yes, but we may want a different hash in this scenario:
```python

@dataclass 
class A:
  fld: int

@dataclass 
class B:
  fld: int


assert stable_hash(A(42)) != stable_hash(B(42))

```

Reply:
Y'all killing me. I'll try pickling tomorrow. 😑

Reply:
who knew this was such an padoras box 😂

Reply:
There is a reason we no longer try to use Python object hashes in persistent cache keys anywhere in the engine... pain in the butt.

Reply:
I expect that always including the type name of the field, and skipping pickling (which involves lots of magic) would be simpler.

Reply:
okie dokie